backend/project/routes_webauthn_duo.py

```diff
@@ -0,0 +1,282 @@
+def verify_credential_info():
+    challenge = session['challenge']
+    email = session['register_email']
+    display_name = session['register_display_name']
+    ukey = session['register_ukey']
+
+    registration_response = request.form
+    trust_anchor_dir = os.path.join(
+        os.path.dirname(os.path.abspath(__file__)), TRUST_ANCHOR_DIR)
+    trusted_attestation_cert_required = True
+    self_attestation_permitted = True
+    none_attestation_permitted = True
+
+    webauthn_registration_response = webauthn.WebAuthnRegistrationResponse(
+        RP_ID,
+        ORIGIN,
+        registration_response,
+        challenge,
+        trust_anchor_dir,
+        trusted_attestation_cert_required,
+        self_attestation_permitted,
+        none_attestation_permitted,
+        uv_required=False)  # User Verification
+
+    try:
+        webauthn_credential = webauthn_registration_response.verify()
+    except Exception as e:
+        return jsonify({'fail': 'Registration failed. Error: {}'.format(e)})
+
+    # Step 17.
+    #
+    # Check that the credentialId is not yet registered to any other user.
+    # If registration is requested for a credential that is already registered
+    # to a different user, the Relying Party SHOULD fail this registration
+    # ceremony, or it MAY decide to accept the registration, e.g. while deleting
+    # the older registration.
+    dbcursor.execute('SELECT credential_id FROM test WHERE credential_id = %s', (webauthn_credential.credential_id,))
+    credential_id_exists = dbcursor.fetchone()
+    
+    if credential_id_exists:
+        return make_response(
+            jsonify({
+                'fail': 'Credential ID already exists.'
+            }), 401)
+
+    dbcursor.execute('SELECT email FROM test WHERE email = %s', (email,))
+    existing_user = dbcursor.fetchone()
+    if not existing_user:
+        if sys.version_info >= (3, 0):
+            webauthn_credential.credential_id = str(
+                webauthn_credential.credential_id, "utf-8")
+            webauthn_credential.public_key = str(
+                webauthn_credential.public_key, "utf-8")
+        '''
+        user = User(
+            ukey=ukey,
+            username=username,
+            display_name=display_name,
+            pub_key=webauthn_credential.public_key,
+            credential_id=webauthn_credential.credential_id,
+            sign_count=webauthn_credential.sign_count,
+            rp_id=RP_ID,
+            icon_url='https://example.com')
+        db.session.add(user)
+        db.session.commit()
+        '''
+        # Add new user entry to database table
+        dbcursor.execute('INSERT INTO users (email, display_name, ukey, icon_url, credential_id, pub_key, sign_count, rp_id) VALUES (%s, %s, %s, %s, %s, %s, %d, %s)', (email, display_name, ukey, 'https://example.com', webauthn_credential.credential_id, webauthn_credential.public_key, webauthn_credential.sign_count, RP_ID))
```

Comment:
One thing that Jonah has been doing is to use stored procedures for our SQL queries instead of doing them directly in our code. The way they work is you send Jonah the SQL query, he makes a stored procedure, then you call something like `dbcursor.call('name of your procedure', args=(email, ...))`. This is mostly to deal with user permission issues when executing SQL queries (i.e., some users not having `INSERT` or `DELETE` privileges).

Reply:
I've gone ahead and replaced some of the raw SQL calls with available stored procedures in 1cc68fcffa18e4d57a40f9101dce42c63012254d; however, at the moment there aren't procedures that match every need, so I've marked those with TODO comments as a reminder and left the code otherwise intact for now.